File: src/ia_local/mcp_server.py
import asyncio
import json
from typing import Any
import httpx
from mcp.server import Server
from mcp.types import Tool, TextContent

# URL do Ollama rodando localmente
OLLAMA_URL = "http://localhost:11434/api/generate"
MODEL_NAME = "tinyllama"

# Criar instância do servidor MCP
app = Server("ia-local-classifier")

async def classify_image_with_ollama(image_description: str) -> dict[str, Any]:
    """
    Classifica uma imagem como gráfico ou não-gráfico usando o modelo local.

    Args:
        image_description: Descrição textual da imagem

    Returns:
        Dicionário com a classificação e justificativa
    """
    # Montar o prompt completo para o modelo
    prompt = f"""Você é um especialista em análise de imagens. Com base na descrição fornecida, classifique se a imagem é um GRÁFICO ou NÃO-GRÁFICO.

Um GRÁFICO inclui:
- Gráficos de linha, barra, pizza, dispersão
- Diagramas estatísticos
- Visualizações de dados
- Infográficos com dados quantitativos
- Tabelas com visualização gráfica

NÃO-GRÁFICO inclui:
- Fotografias normais
- Ilustrações artísticas
- Capturas de tela de texto
- Desenhos sem dados quantitativos
- Logos e ícones

Descrição da imagem: {image_description}

Responda APENAS no seguinte formato JSON (sem texto adicional):
{{
  "classificacao": "GRAFICO" ou "NAO_GRAFICO",
  "confianca": 0.0 a 1.0,
  "justificativa": "breve explicação da decisão"
}}"""

    try:
        # Fazer requisição ao Ollama
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(
                OLLAMA_URL,
                json={
                    "model": MODEL_NAME,
                    "prompt": prompt,
                    "stream": False
                }
            )
            response.raise_for_status()

            # Extrair a resposta do modelo
            result = response.json()
            model_response = result.get("response", "")

            # Tentar parsear o JSON da resposta
            try:
                # Extrair JSON da resposta (o modelo pode adicionar texto extra)
                start_idx = model_response.find("{")
                end_idx = model_response.rfind("}") + 1

                if start_idx != -1 and end_idx > start_idx:
                    json_str = model_response[start_idx:end_idx]
                    classification = json.loads(json_str)
                else:
                    # Fallback se não conseguir parsear
                    classification = {
                        "classificacao": "NAO_GRAFICO",
                        "confianca": 0.5,
                        "justificativa": "Não foi possível processar a resposta do modelo",
                        "resposta_bruta": model_response
                    }
            except json.JSONDecodeError:
                # Fallback se o JSON estiver malformado
                classification = {
                    "classificacao": "NAO_GRAFICO",
                    "confianca": 0.5,
                    "justificativa": "Erro ao parsear resposta do modelo",
                    "resposta_bruta": model_response
                }

            return classification

    except httpx.HTTPError as e:
        return {
            "erro": f"Erro ao conectar com Ollama: {str(e)}",
            "classificacao": "ERRO",
            "confianca": 0.0,
            "justificativa": "Não foi possível conectar ao serviço de IA local"
        }
    except Exception as e:
        return {
            "erro": f"Erro inesperado: {str(e)}",
            "classificacao": "ERRO",
            "confianca": 0.0,
            "justificativa": "Ocorreu um erro durante o processamento"
        }

@app.list_tools()
async def list_tools() -> list[Tool]:
    """
    Lista as ferramentas disponíveis no MCP Server.
    """
    return [
        Tool(
            name="classify_image",
            description=(
                "Classifica uma imagem como GRÁFICO ou NÃO-GRÁFICO usando IA local. "
                "Fornece uma descrição textual da imagem e recebe uma classificação "
                "com nível de confiança e justificativa."
            ),
            inputSchema={
                "type": "object",
                "properties": {
                    "image_description": {
                        "type": "string",
                        "description": (
                            "Descrição detalhada da imagem a ser classificada. "
                            "Inclua elementos visuais, cores, formas, texto visível, "
                            "e qualquer característica relevante."
                        )
                    }
                },
                "required": ["image_description"]
            }
        )
    ]

@app.call_tool()
async def call_tool(name: str, arguments: Any) -> list[TextContent]:
    """
    Executa a ferramenta solicitada.
    """
    if name == "classify_image":
        image_description = arguments.get("image_description", "")

        if not image_description:
            return [
                TextContent(
                    type="text",
                    text="Erro: A descrição da imagem é obrigatória."
                )
            ]

        # Classificar a imagem
        result = await classify_image_with_ollama(image_description)

        # Formatar resultado
        formatted_result = json.dumps(result, indent=2, ensure_ascii=False)

        return [
            TextContent(
                type="text",
                text=f"Resultado da classificação:\n\n{formatted_result}"
            )
        ]

    return [
        TextContent(
            type="text",
            text=f"Erro: Ferramenta '{name}' não encontrada."
        )
    ]

async def main():
    """
    Função principal para iniciar o servidor MCP.
    """
    from mcp.server.stdio import stdio_server

    async with stdio_server() as (read_stream, write_stream):
        await app.run(
            read_stream,
            write_stream,
            app.create_initialization_options()
        )

if __name__ == "__main__":
    asyncio.run(main())

File: backend/services/ia_local/mcp_server.js
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
} from "@modelcontextprotocol/sdk/types.js";

// URL do Ollama rodando localmente
const OLLAMA_URL = "http://localhost:11434/api/generate";
const MODEL_NAME = "llama3:8b";

/**
 * Classifica uma imagem como gráfico ou não-gráfico usando o modelo local.
 *
 * @param {string} imageDescription - Descrição textual da imagem
 * @returns {Promise<Object>} Dicionário com a classificação e justificativa
 */
async function classifyImageWithOllama(imageDescription) {
  // Montar o prompt completo para o modelo
  const prompt = `Você é um especialista em análise de imagens. Com base na descrição fornecida, classifique se a imagem é um GRÁFICO ou NÃO-GRÁFICO.

Um GRÁFICO inclui:
- Gráficos de linha, barra, pizza, dispersão
- Diagramas estatísticos
- Visualizações de dados
- Infográficos com dados quantitativos
- Tabelas com visualização gráfica

NÃO-GRÁFICO inclui:
- Fotografias normais
- Ilustrações artísticas
- Capturas de tela de texto
- Desenhos sem dados quantitativos
- Logos e ícones

Descrição da imagem: ${imageDescription}

Responda APENAS no seguinte formato JSON (sem texto adicional):
{
  "classificacao": "GRAFICO" ou "NAO_GRAFICO",
  "confianca": 0.0 a 1.0,
  "justificativa": "breve explicação da decisão"
}`;

  try {
    // Fazer requisição ao Ollama
    const response = await fetch(OLLAMA_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: MODEL_NAME,
        prompt: prompt,
        stream: false,
      }),
      signal: AbortSignal.timeout(60000), // 60 segundos timeout
    });

    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    // Extrair a resposta do modelo
    const result = await response.json();
    const modelResponse = result.response || "";

    return {
      resposta_bruta: modelResponse,
    };
  } catch (error) {
    return {
      resposta_bruta: "ERROR",
    };
  }
}

// Criar instância do servidor MCP
const server = new Server(
  {
    name: "ia-local-classifier",
    version: "1.0.0",
  },
  {
    capabilities: {
      tools: {},
    },
  }
);

/**
 * Lista as ferramentas disponíveis no MCP Server.
 */
server.setRequestHandler(ListToolsRequestSchema, async () => {
  return {
    tools: [
      {
        name: "classify_image",
        description:
          "Classifica uma imagem como GRÁFICO ou NÃO-GRÁFICO usando IA local. " +
          "Fornece uma descrição textual da imagem e recebe uma classificação " +
          "com nível de confiança e justificativa.",
        inputSchema: {
          type: "object",
          properties: {
            image_description: {
              type: "string",
              description:
                "Descrição detalhada da imagem a ser classificada. " +
                "Inclua elementos visuais, cores, formas, texto visível, " +
                "e qualquer característica relevante.",
            },
          },
          required: ["image_description"],
        },
      },
    ],
  };
});

/**
 * Executa a ferramenta solicitada.
 */
server.setRequestHandler(CallToolRequestSchema, async (request) => {
  const { name, arguments: args } = request.params;

  if (name === "classify_image") {
    const imageDescription = args.image_description || "";

    if (!imageDescription) {
      return {
        content: [
          {
            type: "text",
            text: "Erro: A descrição da imagem é obrigatória.",
          },
        ],
      };
    }

    // Classificar a imagem
    const result = await classifyImageWithOllama(imageDescription);

    // Formatar resultado
    const formattedResult = JSON.stringify(result, null, 2);

    return {
      content: [
        {
          type: "text",
          text: `Resultado da classificação:\n\n${formattedResult}`,
        },
      ],
    };
  }

  return {
    content: [
      {
        type: "text",
        text: `Erro: Ferramenta '${name}' não encontrada.`,
      },
    ],
  };
});

/**
 * Função principal para iniciar o servidor MCP.
 */
async function main() {
  const transport = new StdioServerTransport();
  await server.connect(transport);
  console.error("MCP Server ia-local-classifier iniciado");
}

main().catch((error) => {
  console.error("Erro ao iniciar servidor:", error);
  process.exit(1);
});
